### PR TITLE
Padding free dpo

### PR DESCRIPTION
**Pull request by dame-cell**
# What does this PR do?

New feature #2422 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

For now this is just a draft will be continuing to work on it

### CODE REVIEW

**Comment by dame-cell** [comment]
so the now it works where if we do padding_free =True it will not pad but the tests keep failing due to attention_mask still being there if padding_free =True

**Comment by dame-cell** [comment]
most of the stuff is done just some small stuff left like dealing with list and converting to tensor

**Comment by dame-cell** [comment]
Hey @osanseviero,

The main idea for using padding_free is mostly in place now, but there are still a few things that need to be done. It would be awesome if you could take a look at the code and let me know if there's anything else I should address or add.

File: tests/test_dpo_trainer.py [hunk 1]
```diff
@@ -483,74 +487,62 @@ def test_dpo_trainer_w_dataset_num_proc(self):
...
+            self.assertNotIn("rejected_attention_mask", batch_paddingfree)
+            
+    def test_dpo_trainer_padding_free_training(self):
         with tempfile.TemporaryDirectory() as tmp_dir:
             training_args = DPOConfig(
                 output_dir=tmp_dir,
                 per_device_train_batch_size=2,
                 max_steps=3,
                 remove_unused_columns=False,
-                gradient_accumulation_steps=4,
```
Comment (winglian) [inline tests/test_dpo_trainer.py:519]: Would be good to have tests for this with gradient accumulation too. perhaps using `pytest.mark.parameterize`?

Comment (dame-cell) [reply tests/test_dpo_trainer.py:519]: All right will do so thanks for reviewing 😎

File: trl/trainer/ppo_config.py [hunk 2]
```diff
@@ -53,6 +53,9 @@ class PPOConfig(OnPolicyConfig):
             Discount factor.
```
Comment (qgallouedec) [inline trl/trainer/ppo_config.py:53]: This modif shouldn't be here, right?

Comment (dame-cell) [reply trl/trainer/ppo_config.py:53]: oh my bad i'll fix it right now

**Comment by dame-cell** [comment]
@osanseviero  I think this fixes it sorry for the small mistake I have been making  , thanks for your patience

**Comment by dame-cell** [comment]
still more work to be done not really ready yet

**Comment by dame-cell** [comment]
the padding_free finally can train , it works but 
it's not optimised 
not really clean code yet 
better refactoring of code  

will do this by tommorow

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -80,19 +80,21 @@
...
         return_tensors (`str`, *optional*, defaults to `"pt"`):
             Type of Tensor to return. Only `"pt"` is currently supported.
+        padding_free (`bool`, *optional*, defaults to `False`):
+            If `True`, returns tensors without padding. Otherwise, inputs are padded to the maximum batch length.
 
     Examples:
     ```python
     >>> from trl import PreferenceCollator
-    >>> collator = PreferenceCollator(pad_token_id=0)
+    >>> collator = PreferenceCollator(pad_token_id=0, padding_free=False)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:97]:
```suggestion
    >>> collator = PreferenceCollator(pad_token_id=0)
```

it's the default

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -111,12 +113,22 @@ class PreferenceCollator(DataCollatorMixin):
      'rejected_attention_mask': tensor([[1, 0, 0],
                                         [1, 1, 1]])
     }
+    >>> # Example with padding_free=True
+    >>> collator.padding_free = True
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:117]:
```suggestion
    >>> collator = PreferenceCollator(pad_token_id=0, padding_free=True)
```

**Comment by qgallouedec** [comment]
I'm not sure that having the padding-free logic in the collator is the best option here, since we concatenate prompt and completions in the trainer. Maybe the easiest to have everything inside the `concatenated_forward` method.

**Comment by dame-cell** [comment]
> I'm not sure that having the padding-free logic in the collator is the best option here, since we concatenate prompt and completions in the trainer. Maybe the easiest is to have everything inside the `concatenated_forward` method.

question is since the concatenated_inputs depends on the collator do i just create position ids and all those stuff in the concatenated_forward??

**Comment by qgallouedec** [comment]
Is sound like the easiest way. Am I missing something?

**Comment by dame-cell** [comment]
yes you are right I think this will be done by today  hopefully

**Comment by qgallouedec** [comment]
> depends on the collator

I don't think it depends on the collator. How would it?

**Comment by dame-cell** [comment]
another question we get the inputs ids and such from concatenated_inputs but they are padded in this case so what do we do ?

**Comment by dame-cell** [comment]
So let me get this straight 

the concatenated_Inputs pad our inputs_ids and such and in our case we do not want to pad them right so  what do we do here ? 
do we create the position ids inside the concatenated_inputs  and also add not pad the input ids and such ??

**Comment by qgallouedec** [comment]
Here

https://github.com/huggingface/trl/blob/6d4ed070f1f53a87fb3cff2eb82a56db093bccc6/trl/trainer/dpo_trainer.py#L1115-L1123

After the flushing left, we could remove pad tokens, and add position ids:

```python
# Flush left to reduce the memory usage 
# [[0, 0, x, x, x, x],  ->  [[x, x, x, x], 
#  [0, x, x, x, 0, 0]]       [x, x, x, 0]] 
for i in range(attention_mask.size(0)): 
    first_one_idx = torch.nonzero(attention_mask[i])[0].item() 
    input_ids[i] = torch.roll(input_ids[i], shifts=-first_one_idx) 
    attention_mask[i] = torch.roll(attention_mask[i], shifts=-first_one_idx) 
    loss_mask[i] = torch.roll(loss_mask[i], shifts=-first_one_idx) 

if self.padding_free:
    # Flush left to reduce the memory usage 
    # input =             input =            pos_ids =           input =                    pod_ids =
    # [[x, x, x, x],  ->  [[x, x, x, x], and [[0, 1, 2, 3],  ->  [x, x, x, x, x, x, x]  and [0, 1, 2, 3, 0, 1, 2]
    #  [x, x, x, 0]]       [x, x, x]]         [0, 1, 2]] 

    ... # code here

**Comment by dame-cell** [comment]
all right awesome actually this make more sense 😭

**Comment by dame-cell** [comment]
@osanseviero  before I push my code again I want to benchmark this with padding and padding_free just to show the performance

**Comment by qgallouedec** [comment]
You can push it, no worry we can still refine after

**Comment by dame-cell** [comment]
Thank you for your understanding! I wanted to let you know that I’m a bit tied up today and tomorrow, so I might not be able to push the code right away. I’ll try to get to it as soon as possible, but please feel free to let me know if there’s a hard deadline I should prioritize.

Thanks for your patience!
I'll keep working on it so I'll try to push it by Tommorow if i can

**Comment by qgallouedec** [comment]
No rush on our side :)

**Comment by dame-cell** [comment]
all right so I think this does it I did check if we can train this on a single T4 gpu colab notebook 
now using the examples scripts provided the file `trl/examples/scripts/dpo.p` with a bit of update 
I was able to train a model using the padding_Free =True 
```python 
python trl/examples/scripts/dpo.py \
    --dataset_name trl-lib/ultrafeedback_binarized \
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
    --learning_rate 5.0e-6 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 8 \
    --gradient_checkpointing \
    --logging_steps 1 \
    --output_dir Qwen2-0.5B-DPO \
    --no_remove_unused_columns \
    --use_peft \
    --lora_r 32 \
    --lora_alpha 16
```
without padding_free it kept saying OOM is this normal or what ?

**Comment by dame-cell** [comment]
@osanseviero Just wanted to follow up on this PR and see if there’s any feedback so far. I’m happy to clarify anything or make updates if needed. Let me know whenever you get a chance—thanks so much for your time! 🙌

**Comment by qgallouedec** [comment]
You still need to revert the changes applied to PPO files. And apply pre-commits

File: trl/trainer/dpo_trainer.py [hunk 5]
```diff
@@ -192,6 +191,8 @@ class DPOTrainer(Trainer):
             The function to use to preprocess the logits before computing the metrics.
         peft_config (`dict`, defaults to `None`):
             The PEFT configuration to use for training. If you pass a PEFT configuration, the model will be wrapped in a PEFT model.
+        padding_free (`bool`, defaults to `False`):
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:194]: this should be in DPOConfig

Comment (dame-cell) [reply trl/trainer/dpo_trainer.py]: done I put it in the dpo config

Comment (dame-cell) [reply trl/trainer/dpo_trainer.py]: if we put it in the DPOConfig then do we remove also remove it from here 
```python 

def __init__(
        self,
        model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
        ref_model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
        args: Optional[DPOConfig] = None,
        data_collator: Optional[DataCollator] = None,
        train_dataset: Optional[Dataset] = None,
        eval_dataset: Optional[Union[Dataset, dict[str, Dataset]]] = None,
        processing_class: Optional[
            Union[PreTrainedTokenizerBase, BaseImageProcessor, FeatureExtractionMixin, ProcessorMixin]
        ] = None,
        model_init: Optional[Callable[[], PreTrainedModel]] = None,
        compute_metrics: Optional[Callable[[EvalLoopOutput], dict]] = None,
        callbacks: Optional[list[TrainerCallback]] = None,
        optimizers: tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
        peft_config: Optional[dict] = None,
        padding_free: bool = False,
    ):
        self.padding_free = padding_free
```

Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py]: yes, otherwise it's duplicated

Comment (dame-cell) [reply trl/trainer/dpo_trainer.py]: Done ,also I did it like this which I think is the correct way 

```python 
        self.is_encoder_decoder = model.config.is_encoder_decoder
        self.is_vision_model = model.config.model_type in MODEL_FOR_VISION_2_SEQ_MAPPING_NAMES.keys()
        self.is_peft_model = is_peft_available() and isinstance(model, PeftModel)
        self.model_adapter_name = args.model_adapter_name
        self.ref_adapter_name = args.ref_adapter_name
        self.reference_free = args.reference_free
        self.padding_free = args.padding_free
```

File: tests/test_dpo_trainer.py [hunk 6]
```diff
@@ -472,7 +472,7 @@ def test_dpo_trainer_padding_token_is_none(self):
 
                 trainer.train()
 
-    def test_dpo_trainer_w_dataset_num_proc(self):
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py:475]: why d you remove this test?

Comment (dame-cell) [reply tests/test_dpo_trainer.py]: forgive me I was testing some stuff and forgot to add it back I push the update changes just now

File: trl/trainer/ppo_config.py (hunk 2 above)
Comment (qgallouedec) [reply trl/trainer/ppo_config.py]: You still have modification in ppo files

**Comment by dame-cell** [comment]
This is the metrics for the padding_free training trained on a T4 gpu colab notebook 
configs
```bash
python trl/examples/scripts/dpo.py \
    --dataset_name trl-lib/ultrafeedback_binarized \
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
    --learning_rate 5.0e-6 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 8 \
    --gradient_checkpointing \
    --logging_steps 1 \
    --output_dir Qwen2-0.5B-DPO \
    --no_remove_unused_columns \
    --use_peft \
    --lora_r 32 \
    --lora_alpha 16
```
Screenshot (79)
still testing the not padding_free

**Comment by dame-cell** [comment]
The new push changes some code due to these problems 
nice but there is a few problem
this is training without padding_free
```python
{'loss': 0.6933, 'grad_norm': 28.669958114624023, 'learning_rate': 4.880000000000001e-06, 'rewards/chosen':

 -0.0004301070875953883, 'rewards/rejected': -0.00019319055718369782, 'rewards/accuracies': 0.5, 'rewards/margins': 

-0.0002369165886193514, 'logps/chosen': -359.64996337890625, 'logps/rejected': -218.63882446289062, 'logits/chosen': 

-3.0950357913970947, 'logits/rejected': -2.8118934631347656, 'epoch': 0.02}
```
and training with padding_free
```python
{'loss': 0.6931, 'grad_norm': 197.54409790039062, 'learning_rate': 4.960000000000001e-06, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/chosen': -972.31005859375, 'logps/rejected': -1276.14697265625, 'logits/chosen': -2.9294216632843018, 'logits/rejected': -2.5271308422088623, 'epoch': 0.01}
{'loss': 0.6791, 'grad_norm': 124.33696746826172, 'learning_rate': 4.92e-06, 'rewards/chosen': 0.008736038580536842, 'rewards/rejected': -0.019646836444735527, 'rewards/accuracies': 1.0, 'rewards/margins': 0.02838287316262722, 'logps/chosen': -656.6194458007812, 'logps/rejected': -656.0517578125, 'logits/chosen': -2.9904067516326904, 'logits/rejected': -2.4395949840545654, 'epoch': 0.02}
{'loss': 0.6568, 'grad_norm': 147.94912719726562, 'learning_rate': 4.880000000000001e-06, 'rewards/chosen': 0.020291520282626152, 'rewards/rejected': -0.05392017588019371, 'rewards/accuracies': 1.0, 'rewards/margins': 0.07421170175075531, 'logps/chosen': -653.7568359375, 'logps/rejected': -713.071044921875, 'logits/chosen': -3.067491292953491, 'logits/rejected': -2.60148286819458, 'epoch': 0.02}
{'loss': 0.6229, 'grad_norm': 213.63063049316406, 'learning_rate': 4.84e-06, 'rewards/chosen': 0.02659912034869194, 'rewards/rejected': -0.12024041265249252, 'rewards/accuracies': 1.0, 'rewards/margins': 0.14683951437473297, 'logps/chosen': -1233.0953369140625, 'logps/rejected': -761.375732421875, 'logits/chosen': -2.791830062866211, 'logits/rejected': -2.6060781478881836, 'epoch': 0.03}
```
the grad_norm is so high for padding_free and the reward/accuracies is always 1.0 which is not correct

**Comment by dame-cell** [comment]
I was able to decrease the grad norm significantly by just normalizing the sequence_logps 
```python
{'loss': 0.6931, 'grad_norm': 0.4152209758758545, 'learning_rate': 4.960000000000001e-06, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/chosen': -2.002589702606201, 'logps/rejected': -2.268846035003662, 'logits/chosen': -2.9294216632843018, 'logits/rejected': -2.5271308422088623, 'epoch': 0.01}
{'loss': 0.6931, 'grad_norm': 0.6821983456611633, 'learning_rate': 4.92e-06, 'rewards/chosen': 5.119443085277453e-05, 'rewards/rejected': -9.413660154677927e-05, 'rewards/accuracies': 1.0, 'rewards/margins': 0.0001453310251235962, 'logps/chosen': -2.0501720905303955, 'logps/rejected': -2.262946605682373, 'logits/chosen': -2.990408420562744, 'logits/rejected': -2.439610719680786, 'epoch': 0.02}
{'loss': 0.693, 'grad_norm': 0.5559526681900024, 'learning_rate': 4.880000000000001e-06, 'rewards/chosen': 7.167309377109632e-05, 'rewards/rejected': -0.00018270165310241282, 'rewards/accuracies': 1.0, 'rewards/margins': 0.00025437475414946675, 'logps/chosen': -1.6646722555160522, 'logps/rejected': -2.071936845779419, 'logits/chosen': -3.0675406455993652, 'logits/rejected': -2.601555347442627, 'epoch': 0.02}

```

**Comment by dame-cell** [comment]
@osanseviero 
Train a qwen model for only 10 steps  using both padding_Free =Trrue and padding_free =False  with a batch_size of 1 with no gradient accumuatlion 
GPU peak memory 
| Metric          | Padding Free=False | Padding Free=True |
|-----------------|--------------------|-------------------|
| PEAK MEMORY  | 13.9              | 9.0              |

here is the  lasts steps  metrics when training 
| Metric               | Padding Free=False | Padding Free=True |
|----------------------|--------------------|-------------------|
| Loss                 | 0.6939            | 0.6914           |
| Grad Norm            | 4.1626            | 5.1364           |
| Rewards/Chosen       | -0.00055          | 0.00207          |
| Rewards/Rejected     | 0.00093           | -0.00152         |
| Rewards/Accuracies   | 0.0               | 1.0              |
| Rewards/Margins      | -0.00148          | 0.00359          |
| LogPs/Chosen         | -57.3851          | -57.3589         |
| LogPs/Rejected       | -29.9984          | -27.7588         |
| Logits/Chosen        | -2.8899           | -2.8894          |
| Logits/Rejected      | -2.7367           | -2.3579          |
| Epoch                | 0.01              | 0.01             |

There still appear to be noticeable differences between the Rewards/Chosen and Rewards/Rejected metrics. Despite my efforts to resolve this,I just could not fix it 
with gradient accumulation of 8 you can fit upto 4 batch for padding_free =True

**Comment by qgallouedec** [comment]
(Please stop tagging osanseviero? Unless you've a good reason. He is not involved here, please don't bother him 🙏)

**Comment by qgallouedec** [comment]
I've worked on top of your work to propose a version of padding-free DPO that I've just merged #2520.
Thanks for your contribution!
